RFC: nixos/release: Make aarch64-linux a supported system again

mweinelt opened the pull request:
The aarch64-linux builders on hydra have had a good track in the last year or so and I think it's a good idea to include them in the supported systems again.

This happens in preparation of improving the distributing of the installer ISOs for aarch64-linux systems and advertise them more prominently on the homepage.

###### Description of changes

This is an educated guess of what is needed to improve aarch64-linux support. Ideally we could leave out aarch64-specific channels starting with the 22.11 release cycle.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

mweinelt committed:
nixos/release-small: Add aarch64-linux to supportedSystems

The aarch64-linux builders on hydra have had a good track in the last
year or so and I think it's a good idea to include them in the default
-small jobset.

This happens in preparation of improving the distribution of the
installer ISOs for aarch64-linux systems and advertise them more
prominently on the homepage.

mweinelt committed:
nixos/release-combined: Move aarch64-linux to supportedSystems

The builders have had a good track record in the last year so this is
worth a try.

samueldr reviewed (approved):
Given that the only reason it's not is that [the previous hydra evaluator had trouble with it enabled](https://github.com/NixOS/nixpkgs/commit/1bfe8f189bd25de73340635d4e94673000f273ec), I think it should be done.

The evaluator was re-written since.

mweinelt committed:
nixos/release-combined: Build graphical ISOs for aarch64-linux

mweinelt committed:
installer: enable vmware guest support on x86 only

The vmware guest module asserts on this exact condition, so let's only
enable it on that condition.

mweinelt committed:
installer: enable xe-guest-utilities only on x86

Tries to find a target specific makefile for aarch64 which does not
exist.

vcunat commented:
I think we should unify this with stable releases.  There we have limited support in the main channel and an additional aarch64-linux channel.  I'm not saying it needs to be that way, but I can't see why handle stable and unstable differently.

samueldr commented:
I'm not even sure why there are different jobsets. It causes issues when the updates are out of sync and you want to update the channel for a mixed deployment.

vcunat commented:
22.05 will end relatively soon, so I think we can keep that one as it is.  But for future I believe it would make sense to unify the approaches in unstable and stable (stable e.g. since 22.11, if we decide soon).

As the build farm is now, with on-demand scaling of builders for both `*-linux`, I think it should be fine to include both platforms in the main big channel.  (and hope there won't be difficult regressions specific to aarch64-linux)  But if we go that way, I see no sense in keeping the `nixos-22.11-aarch64` jobset and channel, except perhaps as redirect.  I mean, if we wanted a separate channel in addition to one with both platforms, it would surely be for x86_64, right? (that would open the question of which should be the default, too)

samueldr commented:
> I mean, if we wanted a separate channel in addition to one with both platforms, it would surely be for x86_64, right? (that would open the question of which should be the default, too)

In my opinion, if it was to happen, we'd want the "unqualified" channel to wait for all systems, and as many qualified channel names as we have supported systems. Weird in-betweens with only one arch exposed this way might be confusing for end-users.

Though I don't think there is any need for per-system channels, but I also don't have any data proving it either way.

vcunat commented:
I wouldn't be keen on creating so many jobs(ets), as the central Hydra element isn't very efficient there.  Caching a build from another jobset apparently needs nontrivial work, e.g. after `staging-next` merge it can take several hours to just do that for those roughly 240k jobs we have in the jobsets attached to the `master` branch.

samueldr commented:
I'm not either. We probably don't need them.

I'm mostly saying that if we ended-up needing them, we'd need to have one for each supported arch, no bizarro in-between.

mweinelt commented:
Thinking of the added eval cost alone makes me dizzy.

vcunat commented:
`-small` won't work as suggested in this PR, because some of the channel-critical tests won't evaluate:
```
in job ‘nixos.tests.installer.separateBoot.aarch64-linux’:
error: Non-EFI boot methods are only supported on i686 / x86_64

in job ‘nixos.tests.installer.lvm.aarch64-linux’:
error: Non-EFI boot methods are only supported on i686 / x86_64

in job ‘nixos.tests.installer.simple.aarch64-linux’:
error: Non-EFI boot methods are only supported on i686 / x86_64
```

vcunat commented:
Maybe we should test this branch on Hydra job(s) before merging, so that we don't immediately block the channel(s).

mweinelt commented:
Moved the mentioned tests over to x86_64-linux only for -small. Testing the jobsets before prime-time sgtm.

vcunat commented on nixos/release-small.nix:
```diff
@@ -85,6 +85,8 @@ in rec {
   tested = let
     onSupported = x: map (system: "${x}.${system}") supportedSystems;
+    onSystems = systems: x: map (system: "${x}.${system}")
+        (pkgs.lib.intersectLists systems (supportedSystems ++ limitedSupportedSystems));
```
No `limitedSupportedSystems` in this file so far.

vcunat replied:
So just this line?
```nix
        (pkgs.lib.intersectLists systems supportedSystems);
```

mweinelt replied:
Hah, yeah. I was in a hurry and didn't test this properly. Fix pushed.

mweinelt commented:
> Maybe we should test this branch on Hydra job(s) before merging, so that we don't immediately block the channel(s).

@vcunat What's the hold up here?

vcunat commented:
Oh, I didn't understand that you want it already.  Let's start with small:
https://hydra.nixos.org/jobset/nixos/pr-193600-aarch64-support#tabs-errors

mweinelt commented:
My hope is that we can test drive this early enough so we can get comfortable with it on unstable and then make an informed decision around channel creation time.

misuzu commented on nixos/release-small.nix:
```diff
@@ -83,45 +83,55 @@ in rec {
+      ])
+      (map onSupported [
+        "nixos.dummy"
+        "nixos.iso_minimal"
```
Maybe we should also include [`nixos.tests.boot.uefiCdrom`](https://hydra.nixos.org/build/193876884).

mweinelt replied:
Sounds good to me.

mweinelt committed:
nixos/release-combined: Enable more jobs on aarch64-linux

mweinelt committed:
nixos/release-small: Test uefi cdrom

mweinelt commented:
These tests are currently unconditionally inherited, but not part of the tested set, so they log an eval error, but they don't make the evaluation fail.

```
in job ‘nixos.tests.installer.lvm.aarch64-linux’:
error: Non-EFI boot methods are only supported on i686 / x86_64

in job ‘nixos.tests.installer.separateBoot.aarch64-linux’:
error: Non-EFI boot methods are only supported on i686 / x86_64

in job ‘nixos.tests.installer.simple.aarch64-linux’:
error: Non-EFI boot methods are only supported on i686 / x86_64
```

mweinelt committed:
nixos/tests/chromium: Enable on aarch64-linux

vcunat commented:
I assume that the jobset is just for evaluation checks, so I cancelled the builds.

mweinelt commented:
> `-small` won't work as suggested in this PR, because some of the channel-critical tests won't evaluate:
> 
> ```
> in job ‘nixos.tests.installer.separateBoot.aarch64-linux’:
> error: Non-EFI boot methods are only supported on i686 / x86_64
> 
> in job ‘nixos.tests.installer.lvm.aarch64-linux’:
> error: Non-EFI boot methods are only supported on i686 / x86_64
> 
> in job ‘nixos.tests.installer.simple.aarch64-linux’:
> error: Non-EFI boot methods are only supported on i686 / x86_64
> ```

They are not part of the tested set, so primarily noise. I'm not sure how I could match the system in this attribute set:

https://github.com/NixOS/nixpkgs/blob/a7582289f5df0ae48af4b6d291da2c0684b73d2c/nixos/release-small.nix#L48-L53

I think it could be ok to leave this as is.

> I assume that the jobset is just for evaluation checks, so I cancelled the builds.

Yeah, eval seems to have worked just fine. Is this good enough as is?

vcunat commented:
I can't see anything blocking this.

vcunat reviewed (approved):
I can't see anything blocking this PR.

vcunat commented:
I realized a bit late, but... I think the worst about this PR are concerns about evaluation performance on Hydra.  (I don't consider that a deal-breaker, though.)  The evals including NixOS tests are expensive and while we've apparently solved memory issues, eval now took more than an hour: https://hydra.nixos.org/eval/1783395

vcunat commented:
Actually, on a [quick glance](https://monitoring.nixos.org/grafana/d/hkRCcV0mk/instance-metrics?orgId=1&refresh=30s&from=now-24h&to=now) during performing the evaluation, the `rhea` machine had a few peaks of RAM usage above 95%, so I hope we won't be running into this again.  You know, packages and tests get added continuously, so that by itself makes the requirements grow over time.
